Make a cancelled or closed booking editable so users can add more information after the action has happened in the system

bryangaledxw opened the pull request:
# Changes in this PR

This PR adds the ability to edit closed and cancelled bookings. It first lays the groundwork for both, then adds closed booking editing, then cancelled booking editing.

This PR also sneaks in a change suggested by Elaine, to not autofill the date when first closing or cancelling a booking, though I need to check that this is wanted when Elaine returns.

This PR has E2E tests, but is awaiting API work before they can be run and the branch merged

The first commit in this PR disables automatic shared type generation, for the sake of being able to run integration tests on CI, and should not be merged while this commit is in place

## Screenshots of UI changes

### Before
![localhost_3000_properties_12e41286-021d-4308-9670-d6e98e7228e1_bedspaces_3f624341-9427-4784-ad6f-7a19304024d6_bookings_b3d06305-933b-42ab-913d-42647c821d72 (1)](https://user-images.githubusercontent.com/94137563/226321409-4fdb2345-3c6a-4444-8eb4-67bb5e0779ac.png)

![localhost_3000_properties_12e41286-021d-4308-9670-d6e98e7228e1_bedspaces_3f624341-9427-4784-ad6f-7a19304024d6_bookings_a0abaec7-3f45-415b-8bd7-6b6a80a48df5 (1)](https://user-images.githubusercontent.com/94137563/226321379-1e782513-0507-4178-ba9e-5bd4d11b124f.png)

### After
![localhost_3000_properties_12e41286-021d-4308-9670-d6e98e7228e1_bedspaces_3f624341-9427-4784-ad6f-7a19304024d6_bookings_b3d06305-933b-42ab-913d-42647c821d72](https://user-images.githubusercontent.com/94137563/226321541-bf44661f-b2e2-4a01-8841-35f7e8ce4902.png)

![localhost_3000_properties_12e41286-021d-4308-9670-d6e98e7228e1_bedspaces_3f624341-9427-4784-ad6f-7a19304024d6_bookings_b3d06305-933b-42ab-913d-42647c821d72_mark-as-closed_edit](https://user-images.githubusercontent.com/94137563/226321558-93bc1e91-b6c3-49d2-b01e-2841ffe54249.png)

![localhost_3000_properties_12e41286-021d-4308-9670-d6e98e7228e1_bedspaces_3f624341-9427-4784-ad6f-7a19304024d6_bookings_a0abaec7-3f45-415b-8bd7-6b6a80a48df5](https://user-images.githubusercontent.com/94137563/226321576-cf46005e-51f7-4817-adc6-679a2d674552.png)

![localhost_3000_properties_12e41286-021d-4308-9670-d6e98e7228e1_bedspaces_3f624341-9427-4784-ad6f-7a19304024d6_bookings_a0abaec7-3f45-415b-8bd7-6b6a80a48df5_cancellations_edit](https://user-images.githubusercontent.com/94137563/226321589-e1eda7f7-eec3-4238-a2c5-0e8561d88081.png)


# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

bryangaledxw committed:
Remove date prepopulation from some booking pages

In response to user feedback, we no longer prepopulate the date when cancelling a booking or marking a booking as departed

bryangaledxw committed:
Extract editable partial template from booking departures page

Ahead of adding a page for editing a booking departure, we extract the form contents that will be shared between these two pages. We mirror this in our testing helper classes, by extracting a shared base class that will be used when testing both these pages

bryangaledxw committed:
Extract editable partial template from booking cancellation page

Ahead of adding a page for editing a booking cancellation, we extract the form contents that will be shared between these two pages. We mirror this in our testing helper classes, by extracting a shared base class that will be used when testing both these pages

bryangaledxw committed:
Update booking factory

We update our booking factory to better reflect the contents of Booking objects in real usage, such as a booking in the confirmed state not having a populated cancelled field

bryangaledxw committed:
Add page for editing booking departures

We update our controller and add a new Nunjucks template to display a page for editing departures. At the same time, we add a helper class for testing this page

bryangaledxw committed:
Add page for editing booking cancellations

We update our controller and add a new Nunjucks template to display a page for editing departures. At the same time, we add a helper class for testing this page

bryangaledxw committed:
Make departure editing page reachable

We route our new controller methods, allowing our departure editing page to be integration tested by direct navigation to a URL

bryangaledxw committed:
Make cancellation editing page reachable

We route our new controller methods, allowing our cancellation editing page to be integration tested by direct navigation to a URL

bryangaledxw committed:
Tidy bookingActions helper

Ahead of expanding this function, we refactor it to use a switch statement, to reduce duplication of "if (booking.status ==="

bryangaledxw committed:
Tidy deriveBookingHistory helper

Ahead of expanding this function to support edited cancellations and departures, we change functions such as getPreviousDepartedBookingState so their exits are in the order we expect them to be reached when the function is called multiple times. This will simplify the code we will add later to this function.

We also extract a compareBookingState function, as we will want to compare cancellations and departures, as well as extensions.

bryangaledxw committed:
Add action for editing departed booking

We update the bookingActions helper, allowing the user to reach the departure editing page

bryangaledxw committed:
Add action for editing cancelled booking

We update the bookingActions helper, allowing the user to reach the cancellation editing page

bryangaledxw committed:
Update deriveBookingHistory

We update deriveBookingHistory to support bookings with edited departures, allowing our booking history age to show these bookings correctly

bryangaledxw committed:
Update deriveBookingHistory

We update deriveBookingHistory to support bookings with edited cancellations, allowing our booking history age to show these bookings correctly

bryangaledxw committed:
Sort imports

bryangaledxw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Make a cancelled or closed booking editable so users can add more information after the action has happened in the system](https://trello.com/c/k48zmNoa/866-make-a-cancelled-or-closed-booking-editable-so-users-can-add-more-information-after-the-action-has-happened-in-the-system)

bryangaledxw committed:
Reduce duplication in booking tests

Ahead of expanding integration tests around booking state changes, we extract and use a helper function to set up the needed stubbed state

bryangaledxw committed:
Fix exact helper

We use this helper function to generate a regular expression to exactly match a string. This is useful in our integration tests, but for strings containing characters with special meaning within a regular expression, we were not generating the correct regular expression. We now fix this by escaping the input string before using it to construct the regular expression

bryangaledxw committed:
Add helper methods to Page

To help us when we want to check that our forms are properly prepopulated, we add these helpers to our base Page class

bryangaledxw committed:
Reorganise booking E2E tests

We reorganise our booking E2E tests, ahead of expanding them to cover editing departures and cancellations. To prevent essentially n-squared growth in booking test steps, we reorganise our booking tests, so we have a scenario per possible end state, rather than a scenario per possible intermediate and end state. We also examine the booking history page within the happy-path scenarios, rather than in its own scenarios. And we go back to the booking page after encountering an error when updating its state, ready for if we want to make another attempt at updating the booking

bryangaledxw committed:
E2E test editing departed bookings

bryangaledxw committed:
E2E test editing cancelled bookings

patrickjfl commented on server/testutils/factories/booking.ts:
```diff
@@ -22,40 +22,57 @@ class BookingFactory extends Factory<Booking> {
     })
   }
 
   arrived() {
```
nice - looks like there are other places across the codebase we could roll these out?

bryangaledxw replied:
Where do you have in mind? I'm not sure where you're thinking of?

patrickjfl replied:
ah just that when I do a search there are some tests across the codebase where we're still explicitly setting statuses - maybe we could refactor those to use these methods?

bryangaledxw replied:
The only cases I could see were things like

```
    const departedBooking = bookingFactory.build({
      ...this.booking,
      status: 'departed',
      departureDate: newDeparture.dateTime,
      departure,
    })
```

where we can't really use bookingFactory.departed(). Are there any others I've missed?

patrickjfl replied:
what's the issue with using bookingFactory.departed() here? I mean it doesn't save much space but would it not work?

bryangaledxw replied:
You could do something like:
```
    const departedBooking = bookingFactory.departed().build({
      ...this.booking,
      departureDate: newDeparture.dateTime,
      departure,
    })
```

but then the `departed` status would get overwritten with the existing status from `this.booking`. I could change the signature of `departed` to take in an existing booking and use that, but then I don't think it's obvious which fields it'd take from the passed-in booking, and which it would overwrite

patrickjfl replied:
ah right yeah, ok. That makes sense. Agree, that sounds like it would complicate things and not be worth the benefit.

patrickjfl commented on cypress_shared/utils/booking.ts:
```diff
@@ -0,0 +1,14 @@
+import premisesFactory from '../../server/testutils/factories/premises'
+import roomFactory from '../../server/testutils/factories/room'
+
+export const setupBookingStateStubs = (booking: Booking): { premises: Premises; room: Room } => {
```
nice! another change to be integrated with what's in #236

patrickjfl commented on server/utils/bookingUtils.ts:
```diff
@@ -165,47 +160,45 @@ const getPreviousBookingState = (booking: Booking): Booking => {
   }
 }
 
 const getPreviousArrivedBookingState = (booking: Booking): Booking => {
```
I'm a bit confused about these functions and the naming is not clearing it up for me - is this getting the previous state for an arrived booking? If so maybe getArrivedBookingPreviousState would be better?

bryangaledxw replied:
Yeah, good shout. I've gone with getPredecessorForArrivedBooking, etc, which I think better describes its purpose, and doesn't overload what we mean by 'state' - in compareBookingState, I'm using 'State' as a generic term for the extension/departure/cancellation objects

patrickjfl commented on e2e/tests/booking.feature:
```diff
@@ -5,62 +5,55 @@ Feature: Manage Temporary Accommodation - Booking
         And I create a bedspace with all necessary details
 
-    Scenario: Creating a booking
+    Scenario: Creating a booking, and cancelling while provisional
```
good idea

patrickjfl commented on e2e/tests/booking.feature:
```diff
@@ -11,6 +11,8 @@ Feature: Manage Temporary Accommodation - Booking
         Then I should see a confirmation for my new booking
         And I cancel the booking
         Then I should see the booking with the cancelled status
+        And I edit the cancelled
```
'I edit the cancelled booking'?

bryangaledxw replied:
Thanks!

patrickjfl reviewed (approved):
Very nice work, methodical and lots of good refactoring.

I've made a few minor comments but other than removing the first commit and running the E2E tests, generally looks good.

bryangaledxw committed:
Update naming in deriveBookingHistory

We change getPreviousXBookingState and similar to getPredecessorForXBooking to better describe their purpose

bryangaledxw commented:
I've made some changes now that I can see this running E2E and have rebased with latest main
- The `edit.njk` pages now have the poDetails header (also reflected in test helper class)
- Fixed some typos in the step names in bookings.feature
- The E2E test steps for check errors when changing a booking state now also go back to the booking page. This is needed now that we're doing more in a single E2E test scenario

bryangaledxw committed:
Update shared types

We also make the necessary changes in our booking factory to populate the new required fields on the Booking type
